JavaSourceSet throws IOB when building type information for shaded classes

Discovered while investigating #1672 

https://github.com/apache/curator/blob/5ca31e3520189fac3c0447e3083a76756a535723/pom.xml#L963-L977

![image](https://user-images.githubusercontent.com/22326988/163854553-4ff0a103-8c65-4c73-a8fa-ee803be8b7d8.png)

```log
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default-cli of goal org.openrewrite.maven:rewrite-maven-plugin:4.23.0-SNAPSHOT:run failed: String index out of range: -17
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: -17
    at java.lang.String.substring (String.java:1841)
    at org.openrewrite.java.marker.JavaSourceSet.typeDeclarationFor (JavaSourceSet.java:203)
    at org.openrewrite.java.marker.JavaSourceSet.packagesToTypeDeclarations (JavaSourceSet.java:99)
    at org.openrewrite.java.marker.JavaSourceSet.build (JavaSourceSet.java:75)
    at org.openrewrite.java.Java11Parser.getSourceSet (Java11Parser.java:299)
    at org.openrewrite.java.Java11Parser.parseInputs (Java11Parser.java:199)
    at org.openrewrite.Parser.parse (Parser.java:39)
    at org.openrewrite.maven.MavenMojoProjectParser.listSourceFiles (MavenMojoProjectParser.java:325)
    at org.openrewrite.maven.AbstractRewriteMojo.listResults (AbstractRewriteMojo.java:177)
    at org.openrewrite.maven.RewriteRunMojo.execute (RewriteRunMojo.java:41)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)

```

Hi @traceyyoshima,

Using Java version 11:

add plugin to `hive/pom.xml`
```xml
      <plugin>
        <groupId>org.openrewrite.maven</groupId>
        <artifactId>rewrite-maven-plugin</artifactId>
        <version>4.23.0-SNAPSHOT</version>
        <configuration>
          <activeRecipes>
            <recipe>org.openrewrite.java.cleanup.IsEmptyCallOnCollections</recipe>
          </activeRecipes>
        </configuration>
      </plugin>
```
from the project root invoke rewrite:  `mvn rewrite:run -X`
   
Logging output:
```LOG
[INFO] Parsing Java main files...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Hive 4.0.0-alpha-2-SNAPSHOT:
[INFO] 
[INFO] Hive Storage API ................................... SUCCESS [  2.778 s]
[INFO] Hive ............................................... SUCCESS [01:09 min]
[INFO] Hive Classifications ............................... SUCCESS [  2.367 s]
[INFO] Hive Shims Common .................................. FAILURE [  4.651 s]
[INFO] Hive Shims 0.23 .................................... SKIPPED
...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:21 min
[INFO] Finished at: 2022-04-19T09:37:52-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.openrewrite.maven:rewrite-maven-plugin:4.23.0-SNAPSHOT:run (default-cli) on project hive-shims-common: Execution default-cli of goal org.openrewrite.maven:rewrite-maven-plugin:4.23.0-SNAPSHOT:run failed: String index out of range: -17 -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.openrewrite.maven:rewrite-maven-plugin:4.23.0-SNAPSHOT:run (default-cli) on project hive-shims-common: Execution default-cli of goal org.openrewrite.maven:rewrite-maven-plugin:4.23.0-SNAPSHOT:run failed: String index out of range: -17
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.PluginExecutionException: Execution default-cli of goal org.openrewrite.maven:rewrite-maven-plugin:4.23.0-SNAPSHOT:run failed: String index out of range: -17
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:148)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.StringIndexOutOfBoundsException: String index out of range: -17
    at java.lang.String.substring (String.java:1841)
    at org.openrewrite.java.marker.JavaSourceSet.typeDeclarationFor (JavaSourceSet.java:203)
    at org.openrewrite.java.marker.JavaSourceSet.packagesToTypeDeclarations (JavaSourceSet.java:99)
    at org.openrewrite.java.marker.JavaSourceSet.build (JavaSourceSet.java:75)
    at org.openrewrite.java.Java11Parser.getSourceSet (Java11Parser.java:299)
    at org.openrewrite.java.Java11Parser.parseInputs (Java11Parser.java:199)
    at org.openrewrite.Parser.parse (Parser.java:39)
    at org.openrewrite.maven.MavenMojoProjectParser.listSourceFiles (MavenMojoProjectParser.java:325)
    at org.openrewrite.maven.AbstractRewriteMojo.listResults (AbstractRewriteMojo.java:177)
    at org.openrewrite.maven.RewriteRunMojo.execute (RewriteRunMojo.java:41)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)

```

--- a/rewrite-java/src/main/java/org/openrewrite/java/marker/JavaSourceSet.java
+++ b/rewrite-java/src/main/java/org/openrewrite/java/marker/JavaSourceSet.java
@@ -190,9 +190,7 @@ public class JavaSourceSet implements Marker {
             // The declarable name of "class A$B {}" is "A$B"
             // The declarable name of class B in "class A { class B {}}" is "A.B"
             StringBuilder sb = new StringBuilder();
-            int classNameStartIndex = classInfo.getPackageName().isEmpty() ?
-                    0 :
-                    classInfo.getPackageName().length() + 1;
+            String packageName = classInfo.getPackageName();
             ClassInfoList outerClasses = classInfo.getOuterClasses();
             // Classgraph orders this collection innermost -> outermost, but type names are declared outermost -> innermost
             for (int i = outerClasses.size() - 1; i >= 0; i--) {
@@ -200,9 +198,18 @@ public class JavaSourceSet implements Marker {
                 if (outerClass.isPrivate() || outerClass.isAnonymousInnerClass() || outerClass.isSynthetic()) {
                     return null;
                 }
-                sb.append(outerClass.getName().substring(classNameStartIndex + sb.length())).append(".");
+                String typeName;
+                if (outerClass.getName().contains(packageName)) {
+                    typeName = outerClass.getName();
+                } else {
+                    if (!classInfo.getName().contains(outerClass.getSimpleName())) {
+                        return null;
+                    }
+                    typeName = classInfo.getName().substring(0, classInfo.getName().indexOf(outerClass.getSimpleName()) + outerClass.getSimpleName().length());
+                }
+                sb.append(typeName.substring(packageName.length() + 1 + sb.length())).append(".");
             }
-            String nameFragment = classInfo.getName().substring(classNameStartIndex + sb.length());
+            String nameFragment = classInfo.getName().substring(packageName.length() + 1 + sb.length());
 
             if (isUndeclarable(nameFragment)) {
                 return null;

--- a/rewrite-test/build.gradle.kts
+++ b/rewrite-test/build.gradle.kts
@@ -38,6 +38,8 @@ dependencies {
 
     // configuration generator for service providers
     implementation("com.google.auto.service:auto-service:latest.release")
+
+    implementation("org.apache.hbase:hbase-shaded-client:2.4.11")
 }
 
 tasks.withType<Javadoc> {

--- a/rewrite-test/src/main/kotlin/org/openrewrite/java/marker/JavaSourceSetTest.kt
+++ b/rewrite-test/src/main/kotlin/org/openrewrite/java/marker/JavaSourceSetTest.kt
@@ -27,6 +27,7 @@ import org.openrewrite.java.internal.JavaReflectionTypeMapping
 import org.openrewrite.java.internal.JavaTypeCache
 import org.openrewrite.java.tree.JavaType
 import java.util.*
+import kotlin.streams.toList
 
 interface JavaSourceSetTest {
 
@@ -45,6 +46,17 @@ interface JavaSourceSetTest {
             .classpath
         assertThat(classpath).isNotEmpty
         assertThat(classpath[0]).isInstanceOf(JavaType.ShallowClass::class.java)
+    }
+
+    @Issue("https://github.com/openrewrite/rewrite/issues/1677")
+    @Test
+    fun shadedJar() {
+        val typeCache = JavaTypeCache()
+
+        val shaded = JavaSourceSet.build("test", JavaParser.dependenciesFromClasspath("hbase-shaded-client"), typeCache, false)
+            .classpath.stream().filter {o -> o.fullyQualifiedName.startsWith("org.apache.hadoop.hbase.shaded")}.toList()
+        assertThat(shaded).isNotEmpty
+        assertThat(shaded[0]).isInstanceOf(JavaType.ShallowClass::class.java)
     }
 
     // This test uses a lot of memory and examines a code path that we don't actually take anywhere right now

